src/services/blocks/trace/Trace.ts

```diff
@@ -0,0 +1,666 @@
+import { ApiPromise } from '@polkadot/api';
+import { AccountInfo } from '@polkadot/types/interfaces';
+import { Registry } from '@polkadot/types/types';
+import { stringCamelCase } from '@polkadot/util';
+import BN from 'bn.js';
+import { InternalServerError } from 'http-errors';
+
+import { IOption, isSome } from '../../../types/util';
+import {
+	ActionGroup,
+	BlockTrace,
+	KeyInfo,
+	Operation,
+	PAccountEvent,
+	PActionEvent,
+	PalletKeyInfo,
+	PEvent,
+	Phase,
+	PhaseOther,
+	SpanWithChildren,
+	SpecialKeyInfo,
+	TraceSpan,
+} from './types';
+
+const EMPTY_KEY_INFO = {
+	error: 'key not found',
+};
+
+/**
+ * Known spans
+ */
+const SPANS = {
+	executeBlock: {
+		name: 'execute_block',
+		target: 'frame_executive',
+	},
+	applyExtrinsic: {
+		name: 'apply_extrinsic',
+		target: 'frame_executive',
+	},
+	onInitialize: {
+		name: 'on_initialize',
+		// Targets vary by pallet
+	},
+	onFinalize: {
+		name: 'on_finalize',
+		// Targets vary by pallet
+	},
+};
+
+// In the future this will need a more chain specific solution that creates
+// keys directly from expanded metadata.
+/**
+ * Get all the storage key names based on the ones built into `api`.
+ *
+ * @param api ApiPromise
+ */
+function getKeyNames(api: ApiPromise): Record<string, KeyInfo> {
+	const moduleKeys = Object.keys(api.query).reduce((acc, mod) => {
+		Object.keys(api.query[mod]).forEach((item) => {
+			const queryObj = api.query[mod][item];
+			let key;
+			try {
+				// Slice off '0x' prefix
+				key = queryObj.key().slice(2);
+			} catch {
+				key = queryObj.keyPrefix().slice(2);
+			}
+
+			acc[key] = {
+				module: mod,
+				item,
+			};
+		});
+
+		return acc;
+	}, {} as Record<string, KeyInfo>);
+
+	const wellKnownKeys = {
+		'3a636f6465': {
+			special: ':code',
+		},
+		'3a686561707061676573': {
+			special: ':heappages',
+		},
+		'3a65787472696e7369635f696e646578': {
+			special: ':extrinsic_index',
+		},
+		'3a6368616e6765735f74726965': {
+			special: ':changes_trie',
+		},
+		'3a6368696c645f73746f726167653a': {
+			special: ':child_storage:',
+		},
+	};
+
+	return { ...moduleKeys, ...wellKnownKeys };
+}
+
+/**
+ * Parses spans by
+ * 1) creating a Map of span's `id` to the the span with its children
+ * 2) creating an array with the same spans
+ *
+ * @param spans spans to parse
+ */
+function parseSpans(
+	spans: TraceSpan[]
+): {
+	spansById: Map<number, SpanWithChildren>;
+	parsedSpans: SpanWithChildren[];
+	executeBlockSpanId: number;
+} {
+	// IMPORTANT for downstream logic: Make sure spans are sorted.
+	spans.sort((a, b) => a.id - b.id);
+
+	const spansById = new Map<number, SpanWithChildren>();
+	const parsedSpans = [];
+	let executeBlockSpanId;
+	for (const span of spans) {
+		const spanWithChildren = { ...span, children: [] };
+
+		if (
```

Comment:
Is it expected to have only one span that fulfills this, or by sorting are you looking for the most recent span that fulfills this conditional?

Reply:
Only one span will fill this. Sorting is actually just to make sure the eventual Operations are in the right order.